src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,482 @@
+//
+// Created by charlie sievers on 6/21/18.
+//
+
+#include <mpi.h>
+#include <stdlib.h>
+#include <iostream>
+#include "dynamical_matrix.h"
+#include "atom.h"
+#include "modify.h"
+#include "domain.h"
+#include "comm.h"
+#include "group.h"
+#include "force.h"
+#include "math_extra.h"
+#include "memory.h"
+#include "bond.h"
+#include "angle.h"
+#include "dihedral.h"
+#include "improper.h"
+#include "kspace.h"
+#include "update.h"
+#include "neighbor.h"
+#include "pair.h"
+#include "timer.h"
+#include "finish.h"
+#include "ctime"
+
+
+using namespace LAMMPS_NS;
+enum{REGULAR,ESKM};
+
+/* ---------------------------------------------------------------------- */
+
+DynamicalMatrix::DynamicalMatrix(LAMMPS *lmp) : Pointers(lmp), fp(NULL)
+{
+    external_force_clear = 1;
+}
+
+/* ---------------------------------------------------------------------- */
+
+DynamicalMatrix::~DynamicalMatrix()
+{
+    if (fp && me == 0) fclose(fp);
+    memory->destroy(dynmat);
+    memory->destroy(final_dynmat);
+    fp = NULL;
+}
+
+/* ----------------------------------------------------------------------
+   setup without output or one-time post-init setup
+   flag = 0 = just force calculation
+   flag = 1 = reneighbor and force calculation
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::setup()
+{
+    // setup domain, communication and neighboring
+    // acquire ghosts
+    // build neighbor lists
+    if (triclinic) domain->x2lamda(atom->nlocal);
+    domain->pbc();
+    domain->reset_box();
+    comm->setup();
+    if (neighbor->style) neighbor->setup_bins();
+    comm->exchange();
+    comm->borders();
+    if (triclinic) domain->lamda2x(atom->nlocal+atom->nghost);
+    domain->image_check();
+    domain->box_too_small_check();
+    neighbor->build(1);
+    neighbor->ncalls = 0;
+    neighbor->every = 2;                       // build every this many steps
+    neighbor->delay = 1;
+    neighbor->ago = 0;
+    neighbor->ndanger = 0;
+
+    // compute all forces
+    force_clear();
+    external_force_clear = 0;
+
+    eflag=0;
+    vflag=0;
+    if (pair_compute_flag) force->pair->compute(eflag,vflag);
+    else if (force->pair) force->pair->compute_dummy(eflag,vflag);
+
+    if (atom->molecular) {
+        if (force->bond) force->bond->compute(eflag,vflag);
+        if (force->angle) force->angle->compute(eflag,vflag);
+        if (force->dihedral) force->dihedral->compute(eflag,vflag);
+        if (force->improper) force->improper->compute(eflag,vflag);
+    }
+
+    if (force->kspace) {
+        force->kspace->setup();
+        if (kspace_compute_flag) force->kspace->compute(eflag,vflag);
+        else force->kspace->compute_dummy(eflag,vflag);
+    }
+
+    //modify->setup_pre_reverse(eflag,vflag);
+    if (force->newton) comm->reverse_comm();
+
+    openfile("dynmat.dyn");
```

Comment:
Please avoid using hardcoded filenames. I see, that you are allowing an option to choose a custom file, so just use a default filename, if no custom name is set and **then** open the file. This way you don't lose a file descriptor, or unintentionally create an empty unwanted file.